### PR TITLE
Tests/jax component/training/model updating

### LINKED ISSUE #535: [TEST] Jax Model Updating
### What do you want to test?
Jax model updating components

### Outline of test structure
* Unit tests
* Test components and hooks

### Definition of done
Passing checks, cover all hooks, edge cases considered

### Mandatory checklist before making a PR
* [ ] The success criteria laid down in “Definition of done” are met.
* [ ] Test code is documented - docstrings for methods and classes, static types for arguments.
* [ ] Documentation is updated - README, CONTRIBUTING, or other documentation.

### PR DESCRIPTION
**Pull request by OmaymaMahjoub**
## What?
Test: Unit tests for Jax Model_updating components
Fix: model_updating.py 
## Why?
Components are currently not tested. Needs to be robust for Jax Mava release
## How?
Test: `MAPGMinibatchUpdate` and  `MAPGEpochUpdate` component from `components/jax/training/mode_updating.py`
Fix: `model_update_minibatch` function from `components/jax/training/mode_updating.py` 
          -> Line 133 remove `metrics[agent_key]=agent_metrics` and return `agent_metrics` instead of `metrics` dictionary
## Extra
Close #535

### CODE REVIEW

File: mava/components/jax/training/model_updating.py [hunk 1]
```diff
@@ -113,7 +113,7 @@ def model_update_minibatch(
             )
 
             # Update the networks and optimizors.
-            metrics = {}
+            # metrics = {}
```
Comment (RuanJohn) [inline mava/components/jax/training/model_updating.py:116]: Maybe remove? Or uncomment?

File: mava/components/jax/training/model_updating.py [hunk 2]
```diff
@@ -130,8 +130,8 @@ def model_update_minibatch(
                     gradients[agent_key]
                 )
                 agent_metrics[agent_key]["norm_updates"] = optax.global_norm(updates)
-                metrics[agent_key] = agent_metrics
-            return (params, opt_states), metrics
+                # metrics[agent_key] = agent_metrics
```
Comment (RuanJohn) [inline mava/components/jax/training/model_updating.py:133]: Similarly here

File: mava/components/jax/training/model_updating.py (hunk 1 above)
Comment (OmaymaMahjoub) [reply mava/components/jax/training/model_updating.py:116]: Okay! also I will add the type so we don't have mypy error

Comment (OmaymaMahjoub) [reply mava/components/jax/training/model_updating.py]: Done!

File: mava/components/jax/training/model_updating.py (hunk 2 above)
Comment (OmaymaMahjoub) [reply mava/components/jax/training/model_updating.py]: Also done!

File: mava/components/jax/training/model_updating.py [hunk 3]
```diff
@@ -130,8 +129,7 @@ def model_update_minibatch(
                     gradients[agent_key]
                 )
                 agent_metrics[agent_key]["norm_updates"] = optax.global_norm(updates)
-                metrics[agent_key] = agent_metrics
-            return (params, opt_states), metrics
+            return (params, opt_states), agent_metrics
```
Comment (mmorris44) [inline mava/components/jax/training/model_updating.py:132]: @arnupretorius @KaleabTessera this seems like a MASSIVE bug which has how been fixed. No idea how our systems were even training with this.

Comment (sash-a) [inline mava/components/jax/training/model_updating.py:132]: totally makes sense, have we had a look at what this does to the graphs?

Comment (OmaymaMahjoub) [reply mava/components/jax/training/model_updating.py:132]: Nope not yet, but we agree that we need to remove metric dictionary since it is unnecessary! If you can check with the team in the standup tomorrow the updates of the graph if this will be merged🙏

Comment (RuanJohn) [reply mava/components/jax/training/model_updating.py:132]: I'll do some quick runs to see how/if this affects any graphs.

Comment (RuanJohn) [reply mava/components/jax/training/model_updating.py:132]: It definitely fixes what it is supposed to and cleans up the per agent trainer metric logging on neptune.

File: tests/jax/components/training/model_updating_test.py [hunk 4]
```diff
@@ -0,0 +1,394 @@
...
+
+    Args:
+        mock_trainer: Trainer
+        mock_optimizer: Optimizer
+    """
+    mini_batch_update = MAPGMinibatchUpdate()
+    mini_batch_update.config.optimizer = mock_optimizer
+    mini_batch_update.on_training_utility_fns(trainer=mock_trainer)
+
+    assert mock_trainer.store.optimizer.test == "Done"
```
[the remark this answers is not in the document]
Comment (OmaymaMahjoub) [reply tests/jax/components/training/model_updating_test.py:287]: This to check that `trainer.store.optimizer = self.config.optimizer` (line 83 modelupdate.py) is true since it is just initializing the optimizer and not calling any methods like init and update from it

File: tests/jax/components/training/model_updating_test.py [hunk 5]
```diff
@@ -0,0 +1,394 @@
...
+            )
+        )
+        assert list(mock_trainer.store.opt_states[net_key][1][2]) == list(
+            jax.tree_map(
+                jnp.zeros_like, mock_trainer.store.networks["networks"][net_key].params
+            )
+        )
+        assert mock_trainer.store.opt_states[net_key][2] == optax.EmptyState()
+
+    assert mock_trainer.store.minibatch_update_fn
```
Comment (RuanJohn) [inline tests/jax/components/training/model_updating_test.py:271]: Maybe also assert here that it is callable since it is supposed to be a function.

File: tests/jax/components/training/model_updating_test.py [hunk 6]
```diff
@@ -0,0 +1,394 @@
...
+    mini_batch_update.on_training_utility_fns(trainer=mock_trainer)
+
+    assert mock_trainer.store.optimizer.test == "Done"
+
+    for net_key in mock_trainer.store.networks["networks"].keys():
+        assert mock_trainer.store.opt_states[net_key] == list(
+            mock_trainer.store.networks["networks"][net_key].params
+        )
+
+    assert mock_trainer.store.minibatch_update_fn
```
Comment (RuanJohn) [inline tests/jax/components/training/model_updating_test.py:294]: Again, check that it is callable

File: tests/jax/components/training/model_updating_test.py [hunk 7]
```diff
@@ -0,0 +1,394 @@
...
+        mock_trainer: trainer
+    """
+    mini_epoch_update = MAPGEpochUpdate()
+    mini_epoch_update.on_training_utility_fns(trainer=mock_trainer)
+
+    assert mock_trainer.store.num_epochs == mini_epoch_update.config.num_epochs
+    assert (
+        mock_trainer.store.num_minibatches == mini_epoch_update.config.num_minibatches
+    )
+    assert mock_trainer.store.epoch_update_fn
```
Comment (RuanJohn) [inline tests/jax/components/training/model_updating_test.py:351]: Similarly here the callable assert

**Review by RuanJohn — COMMENTED** [review]
Looks good! Just some very minor requests from my side.

File: tests/jax/components/training/model_updating_test.py [hunk 8]
```diff
@@ -0,0 +1,394 @@
...
+        mock_minibatch_update_fn
+    """
+    mock_trainer = mock_state_and_trainer[1]
+    mini_epoch_update = MAPGEpochUpdate()
+    mini_epoch_update.on_training_utility_fns(trainer=mock_trainer)
+    mock_trainer.store.minibatch_update_fn = mock_minibatch_update_fn
+
+    state = mock_state_and_trainer[0]
+    # update params and opt_state to function in jax.lax.scan function
+    state["params"] = jnp.array([0, 0, 0])
```
[the remark this answers is not in the document]
Comment (OmaymaMahjoub) [reply tests/jax/components/training/model_updating_test.py:371]: I think this will make the code longer since I am going to build another mock because I tried to make only one mock to be adapted to all the functions testings, what do you think if it is a really important to change I will give it a try?

File: tests/jax/components/training/model_updating_test.py [hunk 9]
```diff
@@ -0,0 +1,394 @@
...
+from mava.components.jax.training import Batch
+from mava.components.jax.training.model_updating import (
+    MAPGEpochUpdate,
+    MAPGMinibatchUpdate,
+)
+from mava.systems.jax.trainer import Trainer
+from mava.types import OLT
+
+
+def fake_grad_fn(
```
Comment (KaleabTessera) [inline tests/jax/components/training/model_updating_test.py:35]: Can we call this `fake_ppo_grad_fn`, since not algs use advantage.

File: tests/jax/components/training/model_updating_test.py [hunk 10]
```diff
@@ -0,0 +1,394 @@
...
+        behaviour_log_probs
+        target_values
+        advantages
+        behavior_values
+    Returns:
+        gradient: fake gradient
+        agent_metrics: fake metrics dictionary
+    """
+    gradient = {
+        "agent_0": jnp.array([5.0, 5.0, 5.0]),
```
Comment (KaleabTessera) [inline tests/jax/components/training/model_updating_test.py:58]: Would be nice to use the agent_keys from the actions like line 64. That way, we can use this method for testing different number of agents.

File: tests/jax/components/training/model_updating_test.py (hunk 9 above)
Comment (OmaymaMahjoub) [reply tests/jax/components/training/model_updating_test.py]: Done!

File: tests/jax/components/training/model_updating_test.py (hunk 10 above)
Comment (OmaymaMahjoub) [reply tests/jax/components/training/model_updating_test.py]: Also done!

File: tests/jax/components/training/model_updating_test.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (sash-a) [reply tests/jax/components/training/model_updating_test.py:368]: I think if changing this affects another test then either leave it as is or try get correct asserts for the state defined in the mock in this test. But leaving is probably fine.

**Review by sash-a — APPROVED** [review]
Looks good to me :smile:

**Review by RuanJohn — APPROVED** [review]
Looks good! Thank you @OmaymaMahjoub :)
